### DISCUSSION TITLE
Using eleventy-plugin-lit is it possible to use @material/mwc-button for SSR?

### DISCUSSION THREAD

**Discussion by lmadueno10** [lit/lit#3127] [General]
Every time I try to include import '@material/mwc-button'; in my component...

![image](https://user-images.githubusercontent.com/64743098/178797197-6b626195-79ac-4556-a49b-97ad73a9da73.png)

and then generate the build of the application generates this exception, someone if it is possible to use @material?

![image](https://user-images.githubusercontent.com/64743098/178797415-591c420f-eb2c-4a2c-9947-082df06334be.png)

when the import of @material/mwc-button is commented out if it allows me to generate normally...

![image](https://user-images.githubusercontent.com/64743098/178797986-a855acd0-5ab0-4472-a4c6-f761fec58211.png)

GITHUB: https://github.com/lmadueno10/estasies

**Comment by coder0107git** [comment]
After doing some digging through the errors I found that in the `package.json` of `@material/mwc-button`, `@material/mwc-ripple`, `@material/mwc-base`, `@material/dom`, and `@material/mwc-icon` the line `"type": "module"` is missing (eleventy requires that line must be present in the projects root `package.json` or all javascript module files extensions end in `.mjs`). Also, in `utils.js` from `@material/mwc-base` eleventy encounters `document.addEventListener` which doesn't exist in nodejs.

**Reply by lmadueno10** [reply]
Thank you very much for the answer, that is, I should add in all these @material/mwc-button, @material/mwc-ripple, @material/mwc-base, @material/dom, and @material/mwc-icon the "type": "module"?

**Reply by aomarks** [reply]
Yeah, a PR that adds type:module to all MWC packages should get approved.

As for `document.addEventListener`, I wonder if that could be deferred in `utils.js` instead of always running at module boot up? Otherwise, maybe we can add a no-op shim for `document.addEventListener` to the Lit SSR DOM shim.

**Reply by lmadueno10** [reply]
Thanks for your answer, regarding "As for document.addEventListener, I wonder if that could be deferred in utils.js" which file are you referring to? Is it possible for me to be able to manipulate it so that it does not generate the error?

**Reply by aomarks** [reply]
Sorry I meant a potential change to mwc itself, not something you could do as a workaround.

**Reply by lmadueno10** [reply]
Okay, thank you very much. I have added type: module in package.json of mwc-button, mwc-ripple, mwc-base, dom, mwc-icon...

![image](https://user-images.githubusercontent.com/64743098/179003539-5a65eae1-2d9d-420c-a551-d31f78cd1b93.png)

but I receive the following exception, do you know how I could trace it please? or maybe I'm missing something?
![image](https://user-images.githubusercontent.com/64743098/179003851-c3798e6b-a1c2-44ed-bc5f-bb88f8c604cc.png)

in the log I receive the following
![image](https://user-images.githubusercontent.com/64743098/179004357-f2d07bee-dbf8-4ad9-9bc9-27303d5ae1ab.png)

**Reply by thescientist13** [reply]
Yeah, have also had some challenges with MWC in ESM contexts while testing compatibility for a project I was working on. Some of the experiences Had ty like not having extensions for imports or and assuming processing of CSS files.  
https://github.com/material-components/material-web/issues/2632

A couple examples from this release I tested
https://unpkg.com/browse/@material/mwc-button@0.25.2/mwc-button.js
```js
import { ButtonBase } from './mwc-button-base';
import { styles } from './styles.css';
```

For example, in your screen shot, it looks like the file not found is missing a _.js_ extension.  

For myself, after making some work arounds for the above issues ultimately I think I got stuck at this point though.
https://github.com/microsoft/tslib/issues/143

However, looking at their releases now, they have made some improvement if I look at the latest version of button, so maybe make sure your are on latest?
https://unpkg.com/browse/@material/mwc-button@0.26.1/mwc-button.js

**Reply by lmadueno10** [reply]
Thank you very much for your answer, it helped me a lot. I installed version 0.26.1 which does add "js" to mwc-button-base and styles.css.js.

![image](https://user-images.githubusercontent.com/64743098/179238691-254736dc-b248-4679-a0d5-70d342824ade.png)

![image](https://user-images.githubusercontent.com/64743098/179238852-b7b06cd0-2292-4bcf-a967-20571d986222.png)

But now what you had commented before just came out "Also, in utils.js from @material/mwc-base eleventy encounters document.addEventListener which doesn't exist in nodejs."

![image](https://user-images.githubusercontent.com/64743098/179239210-f06c0ffb-0735-4839-88da-ca9f288f6d35.png)

I was wondering if I should redefine it? or is it another issue? I can't finish understanding. Thank you very much for your support.

**Reply by thescientist13** [reply]
Yeah, it's most likely that MWC is just not SSR "compatible" yet, combined with having any limitations of the environment you're running in, as any tool doing server rendering like Lit+SSR (generally) should define a [scope of compatibility / support](https://github.com/lit/lit/tree/main/packages/labs/ssr#notes-and-limitations) so users can manage expectations.  Sometimes it's just as easy as keeping DOM stuff out of `constructor` / `connectedCallback`, or use `this.shadowRoot` detection for things like hydration, etc.

So, it's a little give and take between the tooling and the components.  Though some of this is trying to be mitigated through the [**Web Components Community Group**](https://github.com/webcomponents-cg) and one if its initiatives for establishing these "Community" Protocols.  There is one on SSR if you want to follow along and / or chime in!  🙌 
https://github.com/webcomponents-cg/community-protocols

Otherwise, getting packages on a path towards ESM is a great start, so anytime you run along these kinds of issues around just resolving modules, it's great to reach out, file issues, bump threads, etc.

**Comment by lmadueno10** [comment]
Hello everyone and again thank you very much for your support, I tried to add type module to ripple and give a settimeout to the load of addEventListener, do you know how I could bypass it? For me it is not a priority to shadowroot the mwc components, but it does not affect the rest of the SSR (for example: ul li). I hope you can support me, thank you very much

![image](https://user-images.githubusercontent.com/64743098/179535106-6e56ac2a-a59d-4c0d-af42-b19a7dfd159d.png)

I already have all the mwc with type: module, but now only this one remains
![image](https://user-images.githubusercontent.com/64743098/179534981-04c9f5c7-fd14-4e95-af2f-d305e5af08e5.png)

**Reply by thescientist13** [reply]
Unless you actually write code like this and shim it in somehow
```js
document = document || {};
document.addEventListener = () => {};
```

then you will keep getting errors like that.  

So you will have to either
1. Shim it
1. [patch package](https://www.google.com/search?client=safari&rls=en&q=patch+package&ie=UTF-8&oe=UTF-8) MWC itself locally for yourself to avoid calling those methods directly without a browser present, e.g.
    ```js
    // we mostly likely running a browser
    if(!process) {
      // do client side only stuff here
    }
1. Or as [suggested earlier](https://github.com/lit/lit/discussions/3127#discussioncomment-3141966) and recommended long term, get an upstream change into MWC itself
